state.c

```diff
@@ -0,0 +1,337 @@
+    linked = signald_new_state(linking, "linking successful, subscribing", "linking_successful", signald_linking_successful, signald_subscribe_after_link);
+    signald_add_transition(linked, "subscribed", subscribed, signald_subscribed, signald_check_proto_version);
+
+    signald_new_state(linking, "error linking", "linking_error", signald_linking_error, NULL);
+
+    /*
+     * Alright, we're subscribed and we've initiated messaging to check the
+     * protocol version.
+     *
+     * What we get back determines the protocol version, and then we request
+     * the contact list.
+     */
+    // If we get a user back, we have the new protocol
+    user = signald_new_state(subscribed, "new protocol detected, get contacts", "user", signald_new_protocol, signald_initialize_contacts);
+    // If we get no user back, we have the new protocol
+    user_not_registered = signald_new_state(subscribed, "new protocol detected, get contacts", "user_not_registered", signald_new_protocol, signald_initialize_contacts);
+    // If we get an error, we have the old protocol
+    unexpected_error = signald_new_state(subscribed, "old protocol detected, get contacts", "unexpected_error", signald_old_protocol, signald_initialize_contacts);
+
+    /*
+     * Once the contact list comes back, we request the group list.
+     */
+    contacts = signald_new_state(user, "got contacts, get groups", "contact_list", signald_received_contact_list, signald_get_groups);
+    signald_add_transition(user_not_registered, "get_contacts", contacts, signald_received_contact_list, signald_get_groups);
+    signald_add_transition(unexpected_error, "get_contacts", contacts, signald_received_contact_list, signald_get_groups);
+
+    /*
+     * Group list has arrived, transition to running state.
+     */
+    running = signald_new_state(contacts, "running", "group_list", signald_received_group_list, NULL);
+
+    /*
+     * Normal even loop.  We transition from "running" back to "running", with
+     * various for the message types.
+     */
+    signald_add_transition(running, "message", running, signald_received_message, NULL);
+    signald_add_transition(running, "group_created", running, signald_get_groups, NULL);
+    signald_add_transition(running, "group_updated", running, signald_get_groups, NULL);
+    signald_add_transition(running, "group_list", running, signald_received_group_list, NULL);
+    signald_add_transition(running, "account_list", running, signald_received_account_list, NULL);
+
+    current = entrypoint;
+}
+
+gboolean
+signald_handle_message(SignaldAccount *sa, JsonObject *obj)
+{
+    const gchar *type = json_object_get_string_member(obj, "type");
+    SignaldStateTransition *transition = g_hash_table_lookup(current->transitions, type);
+
+    purple_debug_info(SIGNALD_PLUGIN_ID,
+                       "Initializing: In state '%s' received message type '%s'\n", 
```

Comment:
Should clean up this log message, the "Initializing: " prefix is no longer necessary (originally I was planning to only use this machine for initialization, but then I realized I could generalize it).